### ISSUE TITLE
Support audio descriptions (WCAG 2)

### ISSUE THREAD

**Issue by Aratramba** [videojs/video.js#1875]
For a government agency we need to provide a WCAG 2.0 compliant videoplayer. Videojs comes close, but we've not been able to come up with a way to provide an audio description inside the videojs player. The only relevant information I found was [this issue from 2011](http://help.videojs.com/discussions/suggestions/34-support-audiodescriptions). Has there been any progress on this since then?

I think basically what we want to do is have a button that lets a user choose either the default audio track or the audio description audio track.

![audio description button](https://cloud.githubusercontent.com/assets/580312/6243762/28d3ce80-b749-11e4-9377-8671dddcb7df.png)

-

> __audio description__
> narration added to the soundtrack to describe important visual details that cannot be understood from the main soundtrack alone
> * Note 1: Audio description of video provides information about actions, characters, scene changes, on-screen text, and other visual content.
> * Note 2: In standard audio description, narration is added during existing pauses in dialogue. (See also extended audio description.)
> * Note 3: Where all of the video information is already provided in existing audio, no additional audio description is necessary.
> * Note 4: Also called "video description" and "descriptive narration."

http://www.w3.org/TR/UNDERSTANDING-WCAG20/media-equiv-audio-desc-only.html

in other words
> Audio description, also referred to as a video description or more precisely called a visual description, is an additional narration track intended primarily for blind and visually impaired consumers of visual media (including television and film, dance, opera, and visual art). It consists of a narrator talking through the presentation, describing what is happening on the screen or stage during the natural pauses in the audio, and sometimes during dialogue if deemed necessary.[1]

http://en.wikipedia.org/wiki/Audio_description

**Comment by heff** [comment]
Good timing, we were just talking to @OwenEdwards from youdescribe.org about this. We unfortunately don't have an immediate solution for this but are going to be exploring the right approach. Would you be willing to help with this?

**Comment by Aratramba** [comment]
well, that depends on what the client / my boss really. I think for now the solution will be to provide an alternative video in a separate link underneath the player. Not the nicest solution though. I'll keep an eye on this repository to see what's going on with the audio descriptions and offer a helping hand/brain when I can. Would this be the thread to watch, or will there be some place else where communication will be taking place?

**Comment by heff** [comment]
For now we'll let this issue represent the descriptions work, and I'll do my best to update it if there's discussion happening elsewhere.

**Comment by mister-ben** [comment]
@Aratramba Just a thought on the workaround - there are a couple of plugins on the wiki for switching between different quality versions of a video, they might be used to switch between 'normal' and audio description versions of the same video for a nicer experience than a link under the player.

**Comment by Aratramba** [comment]
@mister-ben sounds like a plan, thanks!

**Comment by benpdavison** [comment]
Hello Everyone, 
Don't know whether you have solved your problem yet, however based on @mister-ben 's suggestion I did find a workaround. Probably temporary, as I am sure you are looking for a more robust solution than the one I have implemented. I needed a quick solution to implement audio description for a MSc project.
So working of @mister-ben 's suggestion I had a look into the plugins available and how any might offer a solution. Opting to go for Hussnain1's Video.js-HD-Toggle-Plugin. 
As I was using this video (https://youtu.be/ietBYzhb8Dk) which also has a audio described version (https://youtu.be/8u-hpL70Ggk) I was able to simple rename the HD attributes within Hussnain1's code to AD. It was a simple case of find and replace.
Now on the video player, instead of HD it says AD and the clicking of the button switches between the two videos.
As I say, this is a workaround, and I guess a much more robust option would be to use 2 individual audio tracks that you could switch between and sync in time with the video (I know this was my first thought).
Just thought I would share this with you guys, hope it helps in some small way.

Here is the link to Hussnain1's plugin for those interested: https://github.com/Hussnain1/Video.js-HD-Toggle-Plugin

**Comment by Aratramba** [comment]
@benpdavison this only works for mp4 though, right? We need to provide multiple sources, including flash fallback, so this isn't really an option for that use case. 

https://github.com/Hussnain1/Video.js-HD-Toggle-Plugin/blob/master/HD(JavaScript-method).js#L29

**Comment by OwenEdwards** [comment]
@Aratramba you're absolutely right; a more general solution needs to allow as many alternate source formats as are provided in the original `<source>` tags (or in the `data-setup` or JS). I prototyped a possible solution as [Video.js Described Video Selector](https://github.com/OwenEdwards/videojs-described-video-selector), but having multiple `<source>` tags with the same video format within the `<video>` element is non-standard, and might cause issues.

A better solution would use an API/syntax like [Brightcove's playlist plugin for videojs](https://github.com/brightcove/videojs-playlist), where a set of alternate sources are specified in `data-setup` or JS. Then, it's a matter of making sure that the correct alternate source is selected, using the same logic that selected the original source (potentially with Flash fallback), and handling an error if an appropriate alternative isn't specified.

**Comment by pjotrvdh** [comment]
isn't it a good option to have the audio descriptions as an extra audio track that you can select just like the subtitles?

**Comment by OwenEdwards** [comment]
@pjotrvdh that's certainly an option, although there are two ways it can be done; either a substitute audio track (like the Director's Commentary track on a DVD), where the description is pre-mixed with the main audio, or as an additional track which is mixed with the main audio in the browser/player.

Few players implement this (the main one is [OzPlayer](oz-player.com)), partly because the impact on streaming/buffering can be complicated.

**Comment by pjotrvdh** [comment]
@OwenEdwards Then it would be better to have multiple audio tracks. that way it is also possible to use the function for translations

**Comment by OwenEdwards** [comment]
Based on my earlier comment, there is a plugin to switch between two different videos, where the second one adds audio description, at https://github.com/ICTASolutions/videojs-described-video. It has been tested with VideoJS V4.12.x, and the demo includes switching MP4 files or YouTube videos. The syntax supports multiple source files for one video (e.g. MP4 and OGV), as well as different posters for each video. Feel free to test it out, and report any issues.

**Comment by gkatsev** [comment]
@OwenEdwards does the descriptions text track support count towards this?
Also, we do have alternate audio tracks support in videojs now, so, potentially works as well. Though, probably requires more work if the video isn't something like HLS.

**Comment by OwenEdwards** [comment]
The descriptions text track support and the support for alternate audio streams in an HLS mux definitely address this. The plugin I created (which needs updating to v5.x compatibility) also allows allows people who aren't using / can't use HLS to just created two versions of the video, one with Descriptions and one without, as a solution.

So I think we can close this issue. There are other ways of doing description, but those can be opened as a separate issue (e.g. supporting a separate description audio track which is mixed with the main audio in the player, extended descriptions, etc), but I think the support we have fixes this issue now.
